onnx/shape_inference/implementation.cc

```diff
@@ -185,6 +186,30 @@ static void InferShapesImpl(
   std::unordered_map<std::string, const TensorProto*> inputDataByName;
   for (const auto& tp : g->initializer()) {
     inputDataByName[tp.name()] = &tp;
+    
+    // Consider the tensors from the initializer
+    TypeProto *initializerType = new TypeProto();
+    TypeProto_Tensor* initializerTensorType = initializerType->mutable_tensor_type();
+    // set the shape according to the initializer shape info
+    if (tp.dims_size() != 0) {
+      initializerTensorType->set_elem_type(tp.data_type()); // TensorProto_DataType_FLOAT
+      TensorShapeProto* shape = initializerTensorType->mutable_shape();
+      for (int i = 0 ; i < tp.dims_size(); ++i) {
+        shape->add_dim()->set_dim_value(tp.dims(i)); 
+      }
+    }
+    auto iter = valueTypesByName.find(tp.name());
+    // If it already exists in input, check input and initializer is sync
```

Comment:
My understanding was that an input may override the shape of the initializer providing a default value, so when you're doing the shape inferencing you can only use shape info from a constant initializer (an initializer is constant if IR < 4, or if it has no matching input if IR >= 4). @gramalingam do you agree?

Reply:
Roughly, yes. More precisely: we should NOT do the merge below.

Case 1: If X is both an input and an initializer (regardless of IR version), we should use ONLY the shape information provided in the input type; however, it should be the case that the initializer-value should be consistent with the input-type, so it is okay to CHECK for this consistency. For example, it is okay for the input-type to have a shape [M,N] and the initializer to have shape [100, 256]. It is not okay for the input-type to have shape [M,N] and the initializer to have shape [10, 10, 256]. 

Case 2: X is only an initializer and not an input. (Here IR should be >= 4.) We can use the actual shape of the initializer in the inferencing.

Does that sound right?

Reply:
@skottmckay : the one point I am not clear about is your comment that an initializer is a constant if IR < 4. I need to check back to see if this was the conclusion we reached for IR < 4.

Reply:
ONNX Runtime treats initializers as constant if IR version is < 4 as the initializers had to have matching graph inputs, and to not do so would prevent constant folding running. The ONNX spec doesn't require this though, so I agree with your description of Case 1 and Case 2 handling and that an IR check shouldn't be necessary here.